filclient/filclient.go

```diff
@@ -853,221 +857,113 @@ func (fc *FilClient) RetrieveContent(ctx context.Context, miner address.Address,
 		return nil, err
 	}
 
-	pchaddr, err := fc.getPaychWithMinFunds(ctx, minerOwner)
+	// Get the payment channel and create a lane for this retrieval
+	pchAddr, err := fc.getPaychWithMinFunds(ctx, minerOwner)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to get payment channel: %w", err)
 	}
-
-	// Allocate a lane on our payment channel, usually you want a new lane per retrieval
-	lane, err := fc.pchmgr.AllocateLane(pchaddr)
+	pchLane, err := fc.pchmgr.AllocateLane(pchAddr)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to allocate lane: %w", err)
 	}
 
-	// Use the data transfer protocol to propose the retrieval deal
-	sel := shared.AllSelector()
-	var vouch datatransfer.Voucher = proposal
-	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpid, vouch, proposal.PayloadCID, sel)
+	// Submit the retrieval deal proposal to the miner
+	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpID, proposal, proposal.PayloadCID, shared.AllSelector())
 	if err != nil {
 		return nil, err
 	}
 
-	unsub := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
-		if channelState.ChannelID() == chanid {
-			fmt.Printf("[%s] event: %s %d %s %s\n", time.Now().Format("15:04:05"), event.Message, event.Code, datatransfer.Events[event.Code], event.Timestamp.Format("15:04:05"))
-			fmt.Printf("channelstate: %s %s\n", datatransfer.Statuses[channelState.Status()], channelState.Message())
+	// Used to limit prints to the console about received status
+	lastReceivedUpdate := time.Now()
+
+	// The next nonce (incrementing unique ID starting from 0) for the next voucher
+	var nonce uint64 = 0
+
+	// Set up incoming events handler
+
+	// dtRes receives either an error (failure) or nil (success) which is waited
+	// on and handled below before exiting the function
+	dtRes := make(chan error)
+	unsubscribe := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, state datatransfer.ChannelState) {
```

Comment:
So, part of the reason I initially didnt use the subscribe code was because it sends events for *all* data transfers in progress. For the filc case, we dont notice any issues, but in the case of estuary this will be clogged with thousands of other things going on all at the same time.

To get around this, I think we need to filter handling events based on `state.ChannelID == chanid`.

Reply:
Looking at this a bit more, I think theres a race condition (which absolutely exists in the current code too, but its not clear if its actually an issue) where some events for our transfer might fire after we call `OpenPullDataChannel` but before we call subscribe.

This is a little tricky, since in order to do the filtering i mentioned, we need the channel id, so we cant call subscribe first.
Theres no atomic way to subscribe and open a new channel at the same time... How does go-fil-markets do this?

Reply:
Hmm, that's tricky, I wasn't aware. That could be a solution, I'll think about other ways we could solve this more effectively. Might have to come next week.